src/Entity/EntityId.php

```diff
@@ -30,6 +30,37 @@ public function getSerialization() {
 	}
 
 	/**
+	 * Returns the foreign repository name or empty string for local entities.
+	 * @return string
+	 */
+	public function getRepoName() {
+		if ( $this->isForeign() ) {
+			return strtok( $this->serialization, ':' );
+		}
+
+		return '';
+	}
+
+	/**
+	 * @return bool
+	 */
+	public function isForeign() {
+		return strpos( $this->serialization, ':' ) > 0;
+	}
+
+	protected function sanitizeIdSerialization( $id ) {
+		return $this->upcaseLocalId( ltrim( $id, ':' ) );
+
+	}
+
+	private function upcaseLocalId( $id ) {
+		$parts = explode( ':', $id );
+		$parts[count( $parts ) - 1] = strtoupper( end( $parts ) );
```

Comment:
Was it decided to not have (local) IDs with colons in them? I'm not sure if such a thing would be needed for commons or some such.

If there is no such decision, then how about not doing this little code reuse via inheritance thing and simply putting the few lines in both ItemId and PropertyId? Same goes for the test. As far as I remember the reason for why we have this code and the test like this is that we started out with just having EntityId and only later introduced ItemId and PropertyId, ie for legacy reasons and not by design.